### PR TITLE
MelodyParser - dirty song ready

### PR DESCRIPTION
**Pull request by RedSparrowEU**
Improved parser for "dirty" song file - with too many spaces and pipe symbols.

### CODE REVIEW

File: songs/song3_dirty.txt [hunk 1]
```diff
@@ -0,0 +1,31 @@
+define noteA is pause(HT) | note(C4,    E)
+define noteB is pause(HT) | note(C4;  E4, E)
+define noteCE is pause(HT) | note(C4;  E4;  G4, E)
+define noteCH is pause(HT)   | note(C4;  E4  ;G4, E)
+define melodyA is pause(HT)   | note(E4, Q) | note(G4, Q) | note(A4, H)
+define melodyB is pause(HT) | note(A4, Q) | note(B4, Q) | note(E4, H)
+define halfMelA is note(D4, E) | note(C4, E)
+define halfMelB is note(D4, E)  || | note(E4, H)
```
Comment (firemark) [inline songs/song3_dirty.txt:8]: many | is a syntax error.

Comment (RedSparrowEU) [reply songs/song3_dirty.txt:8]: Sure, but parser is prepared even for this...
However if it is critical, should exit with syntax error message... indeed.

Comment (GregSpa) [reply songs/song3_dirty.txt]: However if it is critical should exit with syntax error message...

File: src/MelodyParser.js [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+        melody = this.buildMelody();
+        console.log('Parsing complete...');
+
+    }
+
+    //split to lines and cleanup - rid of unwanted spaces
+    lines(sng) {
+        //reset parser
+        this.defines = [];
```
Comment (firemark) [inline src/MelodyParser.js:31]: it would be nice to discard comment and add another funcion like `reset`

File: src/MelodyParser.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        this.plays = [];
+        this.sounds = [];
+
+        //initial cleaning
+
+        // make single space separator
+        while (sng !== (sng = sng.replace(/  /g, ' '))) {};
```
Comment (firemark) [inline src/MelodyParser.js:45]: you can change to `sng = sng.replace(/\s+/g, ' ')`;

File: src/MelodyParser.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+        sng = sng.replace(/\( /g, '(');
```
Comment (firemark) [inline src/MelodyParser.js:47]: `sng = sng.replace(/\s+([(),;])\s+/g, '$1');`

File: src/MelodyParser.js (hunk 3 above)
Comment (RedSparrowEU) [reply src/MelodyParser.js:45]: I had feeling about this '+'. Thank you! :)

File: src/MelodyParser.js (hunk 4 above)
Comment (RedSparrowEU) [reply src/MelodyParser.js:47]: Beautiful! Seriously. One line instead several.

Comment (firemark) [reply src/MelodyParser.js:47]: example:
```js
> ' ( A ; B , C ; D ) '.replace(/\s+([(),;])\s+/g, '$1');
'(A;B,C;D)'
```

File: src/MelodyParser.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+                case 'define':
+                    this.defines = Object.assign(this.defines, this.parseCommandDefine(comm));
+                    break;
+
+                case 'play':
+                    plays.push(comm);
+                    break;
+
+                default:
+                    console.log(`WARN: Command not recognized or empty line (${l+1}). `)
```
Comment (firemark) [inline src/MelodyParser.js:88]: nice :)

Comment (RedSparrowEU) [reply src/MelodyParser.js:88]: Just as you requested ;)

File: src/MelodyParser.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        }
+    }
+
+    //parse definitions (aka macros)
+    parseCommandDefine(comm) {
+        let defs = [];
+
+        let name = comm[1];
+        for (let i = 3; i < comm.length; i++) {
+            if (comm[i] !== '|') { // ignore pipe separator
```
Comment (firemark) [inline src/MelodyParser.js]: better is ifError pattern:
```js
if (comm[i] === '|') {
  continue; // or return if is a function
}

Comment (RedSparrowEU) [reply src/MelodyParser.js]: Better because?

Comment (firemark) [reply src/MelodyParser.js]: 1. Code is more flat (not nested)
2. Avoid this: https://external-preview.redd.it/I1CShqp2wBguKcuv21m6XRnREKuw8kyAoCavCAGJPBI.jpg?auto=webp&s=819f73b35497078f717311ec509e8224b89b6e71

See: http://wiki.c2.com/?IfOk

Comment (RedSparrowEU) [reply src/MelodyParser.js]: Good point :)

File: src/MelodyParser.js (hunk 4 above)
Comment (RedSparrowEU) [reply src/MelodyParser.js:47]: Unfortunately this formula also doesn't work. Test failed. I stick with original replaces.

File: src/MelodyParser.js (hunk 3 above)
Comment (RedSparrowEU) [reply src/MelodyParser.js:45]: Unfortunately this formula doesn't work. Test failed. I stick with original replaces.

File: src/MelodyParser.js (hunk 4 above)
Comment (firemark) [reply src/MelodyParser.js:47]: :(

**Pull request merged**